plonky2/src/util/serialization/mod.rs

```diff
@@ -302,6 +340,26 @@ pub trait Read {
         ))
     }
 
+    /// Optionally reads a value of type [`MerkleCap`] from `self` with the given `cap_height`.
+    #[inline]
+    fn read_opt_merkle_cap<F, H>(&mut self, cap_height: usize) -> IoResult<Option<MerkleCap<F, H>>>
+    where
+        F: RichField,
+        H: Hasher<F>,
+    {
+        let is_zk = self.read_bool()?;
+        if is_zk {
+            let cap_length = 1 << cap_height;
+            Ok(Some(MerkleCap(
+                (0..cap_length)
+                    .map(|_| self.read_hash::<F, H>())
+                    .collect::<Result<Vec<_>, _>>()?,
+            )))
+        } else {
+            Ok(None)
+        }
+    }
```

Comment:
This method should not be needed. It is used for `read_proof()` which passes `CommonCircuitData` as argument, hence we already know if we need this extra cap or not.